core/unix/signal_linux.c

```diff
@@ -349,6 +349,30 @@ signalfd_thread_exit(dcontext_t *dcontext, thread_sig_info_t *info)
     TABLE_RWLOCK(sigfd_table, write, unlock);
 }
 
+bool
+handle_pre_extended_syscall_sigmasks(dcontext_t *dcontext, kernel_sigset_t *sigmask,
+                                     size_t sizemask)
+{
+    thread_sig_info_t *info = (thread_sig_info_t *)dcontext->signal_field;
+
+    /* XXX i#2311, #3240: We may currently deliver incorrect signals, because the
+     * native sigprocmask the system call may get interrupted by may not be the same
+     * as the native app expects. In addition to this, the p* variants of above syscalls
+     * are not properly emulated w.r.t. their atomicity setting the sigprocmask and
+     * executing the syscall.
+     */
+    info->pre_syscall_app_sigprocmask = info->app_sigblocked;
+    signal_set_mask(dcontext, sigmask);
```

Comment:
This looks like a bug: you set the mask even when failing the syscall.  The failure will skip the syscall and thus skip the post-syscall handlers and thus never restore the mask.  If we're going to fail and skip the syscall we should not make any changes to the emulated state.

Reply:
Given the multiple bugs in handling these failures, maybe a syscall failure should be added to the test: or if not a permanent part of the regression test, please at least do a one-time execution through a failure.

Reply:
Added to test.